File: app/src/main/java/com/example/myapp/Favorites/AdapterFavorites.java
package com.example.myapp.Favorites;

import android.app.Activity;
import android.content.SharedPreferences;
import android.os.Build;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;

import androidx.annotation.NonNull;
import androidx.navigation.Navigation;
import androidx.recyclerview.widget.RecyclerView;

import com.example.myapp.DBService.Location;
import com.example.myapp.DBService.ShamanDao;
import com.example.myapp.MainApp;
import com.example.myapp.R;
import com.example.myapp.WeatherData.WeatherData;
import com.example.myapp.WeatherService.OpenWeatherRetrofit;
import com.google.android.material.imageview.ShapeableImageView;
import com.google.android.material.textview.MaterialTextView;

import org.jetbrains.annotations.NotNull;

import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

import static com.example.myapp.Common.Utils.LOCATION_ARG_COUNTRY;
import static com.example.myapp.Common.Utils.LOCATION_ARG_NAME;
import static com.example.myapp.WeatherService.OpenWeatherRetrofit.APP_ID;
import static com.example.myapp.WeatherService.OpenWeatherRetrofit.BASE_URL;
import static com.example.myapp.WeatherService.OpenWeatherRetrofit.HTTP;
import static com.example.myapp.WeatherService.OpenWeatherRetrofit.HTTPS;

public class AdapterFavorites extends RecyclerView.Adapter<AdapterFavorites.ViewHolder> {

    private ShamanDao shamanDao;
    private List<Location> locations;
    private OpenWeatherRetrofit openWeatherRetrofit;
    private String lang;
    private String units;
    private Activity activity;

    public AdapterFavorites(Activity activity, SharedPreferences sharedPreferences) {
        this.activity = activity;
        shamanDao = MainApp.getInstance().getShamanDao();
        locations = shamanDao.getFavoriteLocations();

        String baseURL;
        if (Build.VERSION.SDK_INT > Build.VERSION_CODES.LOLLIPOP_MR1) {
            baseURL = HTTPS + BASE_URL;
        } else {
            baseURL = HTTP + BASE_URL;
        }

        openWeatherRetrofit = new Retrofit.Builder()
                .baseUrl(baseURL)
                .addConverterFactory(GsonConverterFactory.create())
                .build()
                .create(OpenWeatherRetrofit.class);
        lang = sharedPreferences.getString("pref_lang", "RU");
        units = sharedPreferences.getString("pref_units", "metric");
    }

    @NonNull
    @Override
    public AdapterFavorites.ViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        View v = LayoutInflater.from(viewGroup.getContext())
                .inflate(R.layout.fragment_favorites_item, viewGroup, false);
        return new ViewHolder(v);
    }

    @Override
    public void onBindViewHolder(@NonNull AdapterFavorites.ViewHolder viewHolder, int i) {
        viewHolder.location = locations.get(i);
        viewHolder.requestOpenWeatherRetrofit();
    }

    @Override
    public int getItemCount() { return locations.size(); }

    class ViewHolder extends RecyclerView.ViewHolder {
        private Location location;
        private View view;
        private MaterialTextView txtFavoriteName;
        private MaterialTextView txtFavoriteCountry;
        private MaterialTextView txtTemperature;
        private ShapeableImageView imgWeatherIcon;
        private ImageButton btnFavoriteDelete;

        private ViewHolder(@NonNull View itemView) {
            super(itemView);
            view = itemView;

            txtFavoriteName = itemView.findViewById(R.id.txtFavoriteName);
            txtFavoriteCountry = itemView.findViewById(R.id.txtFavoriteCountry);
            imgWeatherIcon = itemView.findViewById(R.id.imgWeatherIcon);
            txtTemperature = itemView.findViewById(R.id.txtTemperature);
            btnFavoriteDelete = itemView.findViewById(R.id.btnFavoriteDelete);

            view.setOnClickListener((View view) -> {
                Bundle bundle = new Bundle();
                bundle.putCharSequence(LOCATION_ARG_NAME, location.name);
                bundle.putCharSequence(LOCATION_ARG_COUNTRY, location.country);
                Navigation.findNavController(view).navigate(R.id.actionStart, bundle);
            });

            btnFavoriteDelete.setOnClickListener((View view) -> {
                location.favorite = false;
                new Thread(() -> {
                    shamanDao.updateLocation(location);
                }).start();
                locations.remove(location);
                notifyItemRemoved(getAdapterPosition());
            });
        }

        private void requestOpenWeatherRetrofit() {

            new Thread(() -> {
                openWeatherRetrofit.loadWeather(location.name + "," + location.country, APP_ID, lang, units).enqueue(new Callback<WeatherData>() {
                    @Override
                    public void onResponse(Call<WeatherData> call, Response<WeatherData> response) {
                        if (response.isSuccessful() && response.body() != null) {
                            activity.runOnUiThread(() -> {initViewsByGoodResponse(response.body());});
                        } else {
                            activity.runOnUiThread(() -> {initViewsByFailResponse();});
                        }
                    }

                    @Override
                    public void onFailure(Call<WeatherData> call, Throwable t) {
                            activity.runOnUiThread(() -> {initViewsByFailResponse();});
                    }
                });
            }).start();

        }

        private void initViewsByGoodResponse(@NotNull WeatherData wd) {
            wd.setResources(view.getResources());
            txtFavoriteName.setText(wd.getName());
            txtFavoriteCountry.setText(wd.getCountry());
            imgWeatherIcon.setImageResource(wd.getImageResource());
            txtTemperature.setText(wd.getTempString());
        };

        private void initViewsByFailResponse() {
            txtFavoriteName.setText(view.getResources().getString(R.string.not_found_location_name));
            txtFavoriteCountry.setText(view.getResources().getString(R.string.not_found_location_country));
            imgWeatherIcon.setImageResource(R.drawable.ic_report_problem);
            txtTemperature.setText(view.getResources().getString(R.string.not_found_location_temp));
        }
    }
}

File: app/src/main/java/com/chplalex/shaman/Favorites/AdapterFavorites.kt
package com.chplalex.shaman.Favorites

import android.app.Activity
import android.content.SharedPreferences
import android.os.Build
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import androidx.navigation.Navigation
import androidx.recyclerview.widget.RecyclerView
import com.chplalex.shaman.ui.App.Companion.instance
import com.chplalex.shaman.Common.Utils
import com.chplalex.shaman.DBService.Location
import com.chplalex.shaman.DBService.ShamanDao
import com.chplalex.shaman.R
import com.chplalex.shaman.WeatherData.WeatherData
import com.chplalex.shaman.WeatherService.OpenWeatherRetrofit
import com.google.android.material.imageview.ShapeableImageView
import com.google.android.material.textview.MaterialTextView
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

class AdapterFavorites(
    private val activity: Activity,
    sharedPreferences: SharedPreferences
) :
    RecyclerView.Adapter<AdapterFavorites.ViewHolder>() {

    private val shamanDao: ShamanDao = instance.shamanDao
    private val locations: MutableList<Location>
    private val openWeatherRetrofit: OpenWeatherRetrofit
    private val lang: String
    private val units: String

    init {
        locations = shamanDao.favoriteLocations

        val baseURL = if (Build.VERSION.SDK_INT > Build.VERSION_CODES.LOLLIPOP_MR1)
            OpenWeatherRetrofit.HTTPS + OpenWeatherRetrofit.BASE_URL
        else
            OpenWeatherRetrofit.HTTP + OpenWeatherRetrofit.BASE_URL

        openWeatherRetrofit = Retrofit.Builder()
            .baseUrl(baseURL)
            .addConverterFactory(GsonConverterFactory.create())
            .build()
            .create(OpenWeatherRetrofit::class.java)

        lang = sharedPreferences.getString("pref_lang", "RU") ?: "RU"
        units = sharedPreferences.getString("pref_units", "metric") ?: "metric"
    }

    override fun onCreateViewHolder(viewGroup: ViewGroup, viewType: Int) = ViewHolder(
        LayoutInflater.from(viewGroup.context)
            .inflate(R.layout.fragment_favorites_item, viewGroup, false)
    )

    override fun onBindViewHolder(viewHolder: ViewHolder, i: Int) {
        viewHolder.location = locations[i]
        viewHolder.requestOpenWeatherRetrofit()
    }

    override fun getItemCount() = locations.size

    inner class ViewHolder internal constructor(private val view: View) : RecyclerView.ViewHolder(view) {

        var location: Location? = null

        private val txtFavoriteName: MaterialTextView = itemView.findViewById(R.id.txtFavoriteName)
        private val txtFavoriteCountry: MaterialTextView = itemView.findViewById(R.id.txtFavoriteCountry)
        private val txtTemperature: MaterialTextView = itemView.findViewById(R.id.txtTemperature)
        private val imgWeatherIcon: ShapeableImageView = itemView.findViewById(R.id.imgWeatherIcon)
        private val btnFavoriteDelete: ImageButton = itemView.findViewById(R.id.btnFavoriteDelete)

        init {
            view.setOnClickListener { view: View? ->
                val bundle = Bundle()
                location?.let {
                    bundle.putCharSequence(Utils.LOCATION_ARG_NAME, it.name)
                    bundle.putCharSequence(Utils.LOCATION_ARG_COUNTRY, it.country)
                }
                view?.let { Navigation.findNavController(it).navigate(R.id.actionStart, bundle) }
            }
            btnFavoriteDelete.setOnClickListener {
                location?.let {
                    it.favorite = false
                    Thread { shamanDao.updateLocation(it) }.start()
                    locations.remove(it)
                    notifyItemRemoved(adapterPosition)
                }
            }
        }

        private fun initViewsByGoodResponse(wd: WeatherData) {
            wd.setResources(view.resources)
            txtFavoriteName.text = wd.getName()
            txtFavoriteCountry.text = wd.country
            imgWeatherIcon.setImageResource(wd.imageResource)
            txtTemperature.text = wd.tempString
        }

        private fun initViewsByFailResponse() {
            txtFavoriteName.text = view.resources.getString(R.string.not_found_location_name)
            txtFavoriteCountry.text = view.resources.getString(R.string.not_found_location_country)
            imgWeatherIcon.setImageResource(R.drawable.ic_report_problem)
            txtTemperature.text = view.resources.getString(R.string.not_found_location_temp)
        }

        fun requestOpenWeatherRetrofit() = location?.let {
            Thread {
                openWeatherRetrofit.loadWeather(
                    it.name + "," + it.country,
                    OpenWeatherRetrofit.APP_ID,
                    lang,
                    units
                ).enqueue(object : Callback<WeatherData?> {

                    override fun onResponse(call: Call<WeatherData?>, response: Response<WeatherData?>) {
                        if (response.isSuccessful && response.body() != null) {
                            activity.runOnUiThread { initViewsByGoodResponse(response.body()!!) }
                        } else {
                            activity.runOnUiThread { initViewsByFailResponse() }
                        }
                    }

                    override fun onFailure(call: Call<WeatherData?>, t: Throwable) {
                        activity.runOnUiThread { initViewsByFailResponse() }
                    }

                })
            }.start()
        }
    }
}